### DISCUSSION TITLE
Function overloading

### DISCUSSION THREAD

**Discussion by niksy** [voxpelli/types-in-js#16] [Questions] [answered]
It seems there still isn’t possible to fully describe [function overload](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads) with JSDoc.

There were [some](https://github.com/microsoft/TypeScript/issues/25590) [attempts](https://stackoverflow.com/questions/49523358/document-overloaded-function-in-jsdoc), but they either haven’t worked properly or don’t work anymore.

How do you deal with describing functions with different argument types and outputs?

My current half-approach is to `ts-ignore` output and explictly define return type in comments.

**Comment by wooorm** [comment, marked as answer]
Here’s how I get them to work: https://github.com/micromark/micromark/blob/2b1fafb5f3ad0d1c9329ffbc80be1cbd411f5468/packages/micromark/dev/index.js#L12-L30.

**Comment by niksy** [comment]
Oh, I just checked it on my codebase and it works! It’s a nice workaround for sure! Do you know why this works? Is this something related to closure?

**Reply by wooorm** [reply]
`fnA & fnB` merges signatures. In that example they’re typed inline, but you can also use a `@callback` syntax. I think that if you use predefined functions rather then typing them inline, two levels of nesting and no “closure” is fine, but you’d have to check that.
